### PR TITLE
Broker: remove duplicate code in searchForCandidateBroker

### PR DESCRIPTION
**Pull request by letterwuyu**
*(If this PR fixes a github issue, please add `Fixes #<xyz>`.)*

Fixes #<xyz>

*(or if this PR is one task of a github issue, please add `Master Issue: #<xyz>` to link to the master issue.)*

Master Issue: #<xyz>

### Motivation

*Explain here the context, and why you're making that change. What is the problem you're trying to solve.*

### Modifications

*Describe the modifications you've done.*

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API: (yes / no)
  - The schema: (yes / no / don't know)
  - The default values of configurations: (yes / no)
  - The wire protocol: (yes / no)
  - The rest endpoints: (yes / no)
  - The admin cli options: (yes / no)
  - Anything that affects deployment: (yes / no / don't know)

### Documentation

Check the box below and label this PR (if you have committer privilege).

Need to update docs? 

- [ ] `doc-required` 
  
  (If you need help on updating docs, create a doc issue)
  
- [x] `no-need-doc` 
  
  (Please explain why)
  
- [ ] `doc` 
  
  (If this PR contains doc changes)

### CODE REVIEW

**Review by codelipenghui — DISMISSED** [review]
More context to see: https://github.com/apache/pulsar/pull/9240/files#r560645214

Looks like the previous issue has been fixed but the code here does not been cleanup.

**Review by Jason918 — APPROVED** [review]

**Review by eolivelli — APPROVED** [review]
@letterwuyu

can you please update the description ?
it is not clear the intent of this PR.

The change looks good to me, it is only a code clean up

**Comment by letterwuyu** [comment]
> @letterwuyu
> 
> can you please update the description ? it is not clear the intent of this PR.
> 
> The change looks good to me, it is only a code clean up

Sorry for the late reply，I have modified the commit description

**Comment by codelipenghui** [comment]
@letterwuyu Could you please help resolve the conflicts?

**Comment by codelipenghui** [comment]
@315157973 @Jason918 @eolivelli I have merged the master branch to this PR to resolve the conflicts, please help review again.

**Review by gaozhangmin — APPROVED** [review]

**Review by limingnihao — APPROVED** [review]

**Review by congbobo184 — APPROVED** [review]

**Review by Jason918 — APPROVED** [review]

**Review by Shoothzj — APPROVED** [review]

**Review by 315157973 — APPROVED** [review]

**Review by hangc0276 — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java [hunk 1]
```diff
@@ -440,14 +440,6 @@ public boolean registerNamespace(NamespaceName nsname, boolean ensureOwned) thro
     private void searchForCandidateBroker(NamespaceBundle bundle,
                                           CompletableFuture<Optional<LookupResult>> lookupFuture,
                                           LookupOptions options) {
-        if (null == pulsar.getLeaderElectionService()) {
-            LOG.warn("The leader election has not yet been completed! NamespaceBundle[{}]", bundle);
-            lookupFuture.completeExceptionally(
-                    new IllegalStateException("The leader election has not yet been completed!"));
```
Comment (michaeljmarshall) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java:446]: I am concerned that we are changing from an future completed exceptionally to a future completed with `Optional.empty()`. In looking at the code that calls `searchForCandidateBroker`, I can see clear cases where we handle an exception differently than handling an empty option.

Here is one such case: 

https://github.com/apache/pulsar/blob/b80680e5385f3ea0ebdb93ba26d9d316cd67eaa4/pulsar-broker/src/main/java/org/apache/pulsar/broker/web/PulsarWebResource.java#L524-L535

In that method, an exception will result in an exceptionally completed future for `isBundleOwnedByAnyBroker`, but an empty option will return a future completed with `false`. The calling methods for `isBundleOwnedByAnyBroker` appear to branch different based on success/failure.

I think we need to investigate this change in behavior a bit more, possible with some tests.

**Review by michaeljmarshall — CHANGES_REQUESTED** [review]
I have concerns about this behavior change. In looking at the "duplicate code", this method preexisted the code that the PR proposes to replace it with. 

Note that the behavior described as the justification for https://github.com/apache/pulsar/pull/8273 doesn't really make sense, given that it relies on `pulsar.getLeaderElectionService()` returning a null value, even though there is a null check just above the added code.

@merlimat - are you able to provide any insight on this PR in light of https://github.com/apache/pulsar/pull/8273?

**Comment by michaeljmarshall** [comment]
An alternative explanation for how https://github.com/apache/pulsar/pull/8273 observed an NPE is that the `leaderElectionService` is mutable state for the `PulsarService` class and it is not correctly synchronized. That could lead to unexpected behavior where the first check for nullity was false but the actually attempt to use it was true.

**Comment by michaeljmarshall** [comment]
If these concerns have already been addressed, I would like to see someone comment/document that the behavior change is fine. Thus far, I don't see any documentation other than the justification that this PR is "removing duplicate code".

**Comment by tisonkun** [comment]
Closed as stale and conflict. Please rebase and resubmit the patch if it's still relevant.

**Pull request closed without merging**